### DISCUSSION TITLE
How can I publish a Quarto website and blog on Netlify with GH Actions, while using individual `renv` projects per blog post?

### DISCUSSION THREAD

**Discussion by pedrotercero3** [quarto-dev/quarto-cli#11001] [Q&A] [answered]
### Description

The title is pretty convoluted, so let me explain: I have created a Quarto website (here is a reprex: https://github.com/pedrotercero3/quarto-website-test) which also contains a blog. I am using `renv` globally for the website project, and I am also trying to use individual `renv` environments for each of the blog posts. The rationale is to keep the project organized, so that for example I can use version 2 of `ggplot2` in one blog post and version 2.5 in another.

I am using the [GitHub Actions workflow available in the Quarto guide](https://quarto.org/docs/publishing/netlify.html#example-knitr-with-renv) to publish the website on Netlify. It works fine when `renv` is only used for the website project. However, when I initialize a `renv` environment in one of the blog posts, the GitHub actions fails because it is using the global `renv` lock.file instead of the specific one for the blog post (so, for example, if I am using `ggplot2` in the blog post but not on the website, it can't find it).

My question is: is there a way to modify the GH Actions workflow so that it uses the global `renv` environment for the website and the individual `renv` environments for each of the blog posts? Thanks in advance, and sorry if the question is stupid, I find all of this pretty confusing.

**Comment by mcanouil** [comment, marked as answer]
First, Quarto cannot use multiple environments.
Now, using incremental rendering and freeze, you could use renv profiles for example to activate one and rendering one page, etc. for each.
Then you can write your own GitHub action.
Note that the Quarto GitHub Actions does not do anything fancy, you can take a look at the source code.

For short, you have to write your own workflow to activate/render.

**Reply by pedrotercero3** [reply]
Thank you very much for the quick answer. It's good to know that it is not possible to do this with Quarto, at least not out of the box. Maybe my idea doesn't make that much sense in the end? The thing is that if I use R 4.4.2 and `ggplot2` 3.5.1 for a blog post now, and then the current versions of both 4 years from now for a different blog post, I don't see how using `renv` keeps the whole thing reproducible, given that I can't keep using the old versions of the packages for the old blog post. Is there something that I'm missing and there's already a way to address this in Quarto?

**Reply by cscheid** [reply]
The Quarto feature that we built with this kind of usecase in mind is `freeze`, which @mcanouil brought up in his answer: https://quarto.org/docs/projects/code-execution.html#freeze

**Reply by mcanouil** [reply]
The short idea is to activate `freeze`, then something like the code below for each document

```sh
export FILE=path/to/my-document.qmd
Rscript -e 'renv::activate(profile = Sys.getenv("FILE"))'
Rscript -e 'renv::hydrate(sources = Sys.getenv("FILE"), prompt = FALSE)'
Rscript -e 'renv::snapshot(prompt = FALSE)'
quarto render ${FILE}

...

```

For the whole project using the "freezer":
```sh
Rscript -e 'renv::activate(profile = "default")'
quarto render --use-freezer
```

Code obviously not tested.
